src/services/patternMatcher.ts

```diff
@@ -0,0 +1,810 @@
+    export enum PatternMatchKind {
+        Exact,
+        Prefix,
+        Substring,
+        CamelCase
+    }
+
+    // Information about a match made by the pattern matcher between a candidate and the 
+    // search pattern.
+    export interface PatternMatch {
+        // What kind of match this was.  Exact matches are better than prefix matches which are 
+        // better than substring matches which are better than CamelCase matches.
+        kind: PatternMatchKind;
+
+        // If this was a camel case match, how strong the match is.  Higher number means
+        // it was a better match.
+        camelCaseWeight: number;
+
+        // If this was a match where all consituent parts of the candidate and search pattern 
+        // matched case sensitively or case insensitively.  Case sensitive matches of the kind
+        // are better matches than insensitive matches.
+        isCaseSensitive: boolean;
+
+        // Whether or not this match occured with the punctuation from the search pattern stripped
+        // out or not.  Matches without the punctuation stripped are better than ones with punctuation
+        // stripped.
+        punctuationStripped: boolean;
+    }
+
+    // The pattern matcher maintains an internal cache of information as it is used.  Therefore,
+    // you should not keep it around forever and should get and release the matcher appropriately
+    // once you no longer need it.
+    export interface PatternMatcher {
+        // Used to match a candidate against the last segment of a possibly dotted pattern.  This
+        // is useful as a quick check to prevent having to compute a container before calling 
+        // "getMatches".
+        //
+        // For example, if the search pattern is "ts.c.SK" and the candidate is "SyntaxKind", then
+        // this will return a successful match, having only tested "SK" against "SyntaxKind".  At
+        // that point a call can be made to 'getMatches("SyntaxKind", "ts.compiler")', with the
+        // work to create 'ts.compiler' only being done once the first match succeeded.
+        getMatchesForLastSegmentOfPattern(candidate: string): PatternMatch[];
+
+        // Fully checks a candidate, with an optional dotted container against the search pattern.
+        // The candidate must match the last part of the search pattern, and the dotted container
+        // must match the preceding segments of the pattern.
+        getMatches(candidate: string, dottedContainer: string): PatternMatch[];
+        isDottedPattern: boolean;
+    }
+
+    // First we break up the pattern given by dots.  Each portion of the pattern between the
+    // dots is a 'Segment'.  The 'Segment' contains information about the entire section of 
+    // text between the dots, as well as information about any individual 'Words' that we 
+    // can break the segment into.  A 'Word' is simply a contiguous sequence of characters

```

Comment:
Seems like if you have segments separated by dots, you are not typing something that can have spaces interspersed.